sssd_test_framework/utils/dbus_types.py

```diff
@@ -0,0 +1,675 @@
+        return f"{self._type_prefix}:{self._element.param()}"
+
+
+class DBUSTypeDict(DBUSTypeContainer):
+
+    _type_prefix = "dict"
+
+    def __init__(self, model_key: DBUSType | None = None, model_value: DBUSType | None = None):
+        """
+        Build a dictionary containing elements of a single type and keys of
+        another type.
+
+        An object of each type can be passed as parameter. This object will be
+        cloned to create the elements when values are assigned.
+
+        If the objects are not passed, we will try to guess them. This behavior
+        is required because the variant type doesn't include the subtype in
+        the signature.
+        """
+        if (model_key is None or model_value is None) and model_key != model_value:
+            raise ValueError("Both arguments or none must be None")
+
+        self._model_key = model_key
+        self._model_value = model_value
+        self._str: str | None = None
+
+    @property
+    def value(self) -> dict:
+        return dict(self._value)
+
+    @value.setter
+    def value(self, val: dict):
+        if self._model_key is None or self._model_value is None:
+            raise ValueError("No subtype set")
+
+        self._value = dict(val)
+        self._dict = {}
+        for k in val.keys():
+            ko = self._model_key.clone()
+            ko.value = k
+            vo = self._model_value.clone()
+            vo.value = val[k]
+            self._dict[ko] = vo
+
+    def __str__(self) -> str:
+        if self._str is None:
+            val = ""
+            delim = ""
+            for k, v in self._dict.items():
+                val += f"{delim}{k},{v}"
+                delim = ","
+            self._str = val
+
+        return self._str
+
+    def clone(self):
+        return self.__class__(self._model_key, self._model_value)
+
+    # We may need to handle dict entry as a separate type as that's how D-Bus
+    # considers it.
+    # For the moment, there is no gain in doing it, so we keep this.
+    def _parse_dict_entry(self, result: DBUSResult) -> tuple:
+        res = re.match(r"^ *dict entry *\(", result.text)
+        if res is None:
+            raise TypeError("Not a dictionary entry")
+
+        result.text = result.text.removeprefix(res.group(0))
+
+        if self._model_key is None:
+            self._model_key = DBUSType._guess_next_type(DBUSResult(result.text))
+        key = self._model_key.clone()
+        key.parse(result)
+
+        if self._model_value is None:
+            self._model_value = DBUSType._guess_next_type(DBUSResult(result.text))
+        value = self._model_value.clone()
+        value.parse(result)
+
+        res = re.match(r"^ *\) *", result.text)
+        if res is None:
+            raise TypeError("Malformed dictionary entry: unfinished expression")
+
+        result.text = result.text.removeprefix(res.group(0)).strip()
+        return key, value
+
+    # Parsing: (without the newlines)
```

Comment:
If there is a newline, matching may fail depending on the case.

re.MULTILINE could help in this case, but I still need to remove the new lines as the expressions are consumed. So instead of doing it all at once, I would be doing it bit by bit while processing the expressions. Considering this is for internal use of the framework (users should not call `parse()` themselves), I prefer to keep it like it is.